cmd/sync-rover-groups/ldap.go

```diff
@@ -0,0 +1,80 @@
+package main
+
+import (
+	"fmt"
+	"strings"
+
+	ldapv3 "github.com/go-ldap/ldap/v3"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+type ldapGroupResolver struct {
+	conn ldapConn
+}
+
+type ldapConn interface {
+	Search(searchRequest *ldapv3.SearchRequest) (*ldapv3.SearchResult, error)
+}
+
+const (
+	baseDN = "dc=redhat,dc=com"
+)
+
+type notFoundErr struct {
+	group string
+}
+
+// IsNotFoundError returns true if the error indicates the provided
+// object is not found.
+func IsNotFoundError(err error) bool {
+	if err == nil {
+		return false
+	}
+	_, ok := err.(*notFoundErr)
+	return ok
+}
+
+func (e *notFoundErr) Error() string {
+	return fmt.Sprintf("failed to find group on the ldap server: %s", e.group)
+}
+
+func (r *ldapGroupResolver) resolve(name string) (*Group, error) {
+	filter := fmt.Sprintf("(&(objectClass=rhatRoverGroup)(cn=%s))", ldapv3.EscapeFilter(name))
+	searchReq := ldapv3.NewSearchRequest(baseDN, ldapv3.ScopeWholeSubtree, 0, 0, 0, false, filter, []string{"uniqueMember"}, []ldapv3.Control{})
+
+	result, err := r.conn.Search(searchReq)
```

Comment:
Can `r.conn` be nil? If yes we should check to avoid nil poitner dereference. Even if this is checked in another place, it is better to be defensive and panic with an explicit error ("connection to LDAP is nil") rather than NPD. You'd guard against a potential bug.

Reply:
Updated.